atomicapp/nulecule/config.py

```diff
@@ -0,0 +1,149 @@
+
+    Data sources:
+        cli: Config data coming from the CLI
+        runtime: Config data resolved during atomic app runtime. For example,
+            when the value for a parameter in a Nulecule or Nulecule graph
+            item is missing in answers data, we first try to load the default
+            value for the parameter. When there's no default value, or when
+            the user has specified to forcefully ask the user for values, we
+            ask the user for data. These data collected/resolved during runtime
+            form the runtime data.
+        answers: Config data coming from answers file
+        defaults: Default config data specified in atomicapp/constants.py
+
+    The priority order of the data sources is:
+    cli > runtime > answers > defaults
+    """
+
+    PRIORITY = (
+        'cli',
+        'runtime',
+        'answers',
+        'defaults'
+    )
+
+    def __init__(self, answers=None, cli=None):
+        """
+        Initialize a Config instance.
+
+        Args:
+            answers (dict): Answers data
+            cli (dict): CLI data
+        """
+        answers = answers or {}
+        cli = cli or {}
+        # We use a defaultdict of defaultdicts so that we can avoid doing
+        # redundant checks in a nested dictionary if the value of the keys
+        # are dictionaries or None.
+        self._data = defaultdict(defaultdict)
+        # Initialize default data dict
+        self._data['defaults'] = defaultdict(defaultdict)
+        # Initialize answers data dict
+        self._data['answers'] = defaultdict(defaultdict)
+        # Initialize cli data dict
+        self._data['cli'] = defaultdict(defaultdict)
+        # Initialize runtime data dict
+        self._data['runtime'] = defaultdict(defaultdict)
+
+        # Load default answers
+        for scope, data in DEFAULT_ANSWERS.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='defaults')
+        self.set('provider', DEFAULT_PROVIDER, scope=GLOBAL_CONF, source='defaults')
+
+        # Load answers data
+        for scope, data in answers.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='answers')
+
+        # Load cli data
+        for key, value in cli.items():
+            self.set(key, value, scope=GLOBAL_CONF, source='cli')
+
+    def get(self, key, scope=GLOBAL_CONF):
+        """
+        Get the value of a key in a scope. This takes care of resolving
+        the value by going through the PRIORITY order of the various
+        sources of data.
+
+        Args:
+            key (str): Key
+            scope (str): Scope from which to fetch the value for the key
+
+        Returns:
+            Value for the key.
+        """
+        for source in self.PRIORITY:
+            value = self._data[source][scope].get(key)
+            if value:
+                return value
```

Comment:
I know `None` will get returned here if no call to return is given but should we explicitly do it anyway in order to stop people being confused about whether we meant to do it that way or not.